File: apps/php/src/LinkedList.php
<?php

declare(strict_types=1);

namespace Algorithm;

/**
 * 第8章 リスト
 */

// ---- 線形リスト（単方向） ----

class Node
{
    public function __construct(
        public mixed $data,
        public ?Node $next = null,
    ) {}
}

class LinkedList
{
    private ?Node $head = null;
    private int   $size = 0;

    public function isEmpty(): bool
    {
        return $this->head === null;
    }

    public function length(): int
    {
        return $this->size;
    }

    public function addFirst(mixed $data): void
    {
        $this->head = new Node($data, $this->head);
        $this->size++;
    }

    public function addLast(mixed $data): void
    {
        $node = new Node($data);
        if ($this->head === null) {
            $this->head = $node;
        } else {
            $cur = $this->head;
            while ($cur->next !== null) {
                $cur = $cur->next;
            }
            $cur->next = $node;
        }
        $this->size++;
    }

    public function removeFirst(): void
    {
        if ($this->head === null) {
            throw new \UnderflowException('List is empty');
        }
        $this->head = $this->head->next;
        $this->size--;
    }

    public function removeLast(): void
    {
        if ($this->head === null) {
            throw new \UnderflowException('List is empty');
        }
        if ($this->head->next === null) {
            $this->head = null;
        } else {
            $cur = $this->head;
            while ($cur->next->next !== null) {
                $cur = $cur->next;
            }
            $cur->next = null;
        }
        $this->size--;
    }

    /** ノード $target を削除 */
    public function remove(Node $target): void
    {
        if ($this->head === null) {
            return;
        }
        if ($this->head === $target) {
            $this->head = $this->head->next;
            $this->size--;
            return;
        }
        $cur = $this->head;
        while ($cur->next !== null) {
            if ($cur->next === $target) {
                $cur->next = $cur->next->next;
                $this->size--;
                return;
            }
            $cur = $cur->next;
        }
    }

    public function search(mixed $data): ?Node
    {
        $cur = $this->head;
        while ($cur !== null) {
            if ($cur->data === $data) {
                return $cur;
            }
            $cur = $cur->next;
        }
        return null;
    }

    public function includes(mixed $data): bool
    {
        return $this->search($data) !== null;
    }

    public function clear(): void
    {
        $this->head = null;
        $this->size = 0;
    }
}

// ---- 双方向リスト ----

class DNode
{
    public function __construct(
        public mixed  $data,
        public ?DNode $prev = null,
        public ?DNode $next = null,
    ) {}
}

class DoublyLinkedList
{
    private ?DNode $head = null;
    private ?DNode $tail = null;
    private int    $size = 0;

    public function isEmpty(): bool
    {
        return $this->head === null;
    }

    public function length(): int
    {
        return $this->size;
    }

    public function addFirst(mixed $data): void
    {
        $node = new DNode($data, null, $this->head);
        if ($this->head !== null) {
            $this->head->prev = $node;
        } else {
            $this->tail = $node;
        }
        $this->head = $node;
        $this->size++;
    }

    public function addLast(mixed $data): void
    {
        $node = new DNode($data, $this->tail, null);
        if ($this->tail !== null) {
            $this->tail->next = $node;
        } else {
            $this->head = $node;
        }
        $this->tail = $node;
        $this->size++;
    }

    public function remove(DNode $target): void
    {
        if ($target->prev !== null) {
            $target->prev->next = $target->next;
        } else {
            $this->head = $target->next;
        }
        if ($target->next !== null) {
            $target->next->prev = $target->prev;
        } else {
            $this->tail = $target->prev;
        }
        $this->size--;
    }

    public function search(mixed $data): ?DNode
    {
        $cur = $this->head;
        while ($cur !== null) {
            if ($cur->data === $data) {
                return $cur;
            }
            $cur = $cur->next;
        }
        return null;
    }

    public function includes(mixed $data): bool
    {
        return $this->search($data) !== null;
    }

    public function clear(): void
    {
        $this->head = null;
        $this->tail = null;
        $this->size = 0;
    }
}

// ---- 配列によるリスト ----

class ArrayLinkedList
{
    public const NULL_INDEX = -1;

    private array $data;
    private array $next;
    private array $dnext; // フリーリスト用
    private int   $head;
    private int   $deleted;
    private int   $size;
    private int   $max;

    public function __construct(private int $capacity)
    {
        $this->data    = array_fill(0, $capacity, null);
        $this->next    = array_fill(0, $capacity, self::NULL_INDEX);
        $this->head    = self::NULL_INDEX;
        $this->deleted = self::NULL_INDEX;
        $this->size    = 0;
        $this->max     = 0;
    }

    private function alloc(): int
    {
        if ($this->deleted !== self::NULL_INDEX) {
            $idx           = $this->deleted;
            $this->deleted = $this->next[$idx];
            return $idx;
        }
        return $this->max++;
    }

    private function free(int $idx): void
    {
        $this->next[$idx] = $this->deleted;
        $this->deleted    = $idx;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function length(): int
    {
        return $this->size;
    }

    public function addFirst(mixed $data): void
    {
        $idx              = $this->alloc();
        $this->data[$idx] = $data;
        $this->next[$idx] = $this->head;
        $this->head       = $idx;
        $this->size++;
    }

    public function addLast(mixed $data): void
    {
        $idx              = $this->alloc();
        $this->data[$idx] = $data;
        $this->next[$idx] = self::NULL_INDEX;
        if ($this->head === self::NULL_INDEX) {
            $this->head = $idx;
        } else {
            $cur = $this->head;
            while ($this->next[$cur] !== self::NULL_INDEX) {
                $cur = $this->next[$cur];
            }
            $this->next[$cur] = $idx;
        }
        $this->size++;
    }

    public function removeFirst(): void
    {
        if ($this->head === self::NULL_INDEX) {
            throw new \UnderflowException('List is empty');
        }
        $idx        = $this->head;
        $this->head = $this->next[$idx];
        $this->free($idx);
        $this->size--;
    }

    /** データを探してインデックスを返す（なければ NULL_INDEX） */
    public function search(mixed $data): int
    {
        $cur = $this->head;
        while ($cur !== self::NULL_INDEX) {
            if ($this->data[$cur] === $data) {
                return $cur;
            }
            $cur = $this->next[$cur];
        }
        return self::NULL_INDEX;
    }

    public function includes(mixed $data): bool
    {
        return $this->search($data) !== self::NULL_INDEX;
    }
}

File: apps/ruby/lib/algorithm/linked_list.rb
# frozen_string_literal: true

# 第8章 リスト（連結リスト）

module Algorithm
  # 単方向連結リストのノード
  class LLNode
    attr_accessor :data, :next_node

    def initialize(data, next_node = nil)
      @data = data
      @next_node = next_node
    end
  end

  # 線形リスト（単方向連結リスト）
  class LinkedList
    class Empty < StandardError; end

    def initialize
      @head = nil
      @no = 0
    end

    def length
      @no
    end

    def include?(data)
      !search(data).nil?
    end

    def each
      ptr = @head
      while ptr
        yield ptr
        ptr = ptr.next_node
      end
    end

    def empty?
      @head.nil?
    end

    def search(data)
      ptr = @head
      while ptr
        return ptr if ptr.data == data

        ptr = ptr.next_node
      end
      nil
    end

    def add_first(data)
      @head = LLNode.new(data, @head)
      @no += 1
    end

    def add_last(data)
      if @head.nil?
        @head = LLNode.new(data)
      else
        ptr = @head
        ptr = ptr.next_node while ptr.next_node
        ptr.next_node = LLNode.new(data)
      end
      @no += 1
    end

    def remove_first
      raise Empty if @head.nil?

      @head = @head.next_node
      @no -= 1
    end

    def remove_last
      raise Empty if @head.nil?

      if @head.next_node.nil?
        @head = nil
      else
        ptr = @head
        ptr = ptr.next_node while ptr.next_node && ptr.next_node.next_node
        ptr.next_node = nil
      end
      @no -= 1
    end

    def remove(node)
      return if @head.nil?

      if @head.equal?(node)
        @head = @head.next_node
        @no -= 1
        return
      end
      ptr = @head
      while ptr.next_node
        if ptr.next_node.equal?(node)
          ptr.next_node = node.next_node
          @no -= 1
          return
        end
        ptr = ptr.next_node
      end
    end

    def clear
      @head = nil
      @no = 0
    end
  end

  # 双方向連結リストのノード
  class DLLNode
    attr_accessor :data, :prev_node, :next_node

    def initialize(data = nil, prev_node = nil, next_node = nil)
      @data = data
      @prev_node = prev_node
      @next_node = next_node
    end
  end

  # 双方向連結リスト（番兵ノード使用）
  class DoublyLinkedList
    def initialize
      # 番兵ノード: head.next_node が先頭、head.prev_node が末尾
      @head = DLLNode.new
      @head.prev_node = @head
      @head.next_node = @head
      @no = 0
    end

    def length
      @no
    end

    def include?(data)
      !search(data).nil?
    end

    def each
      ptr = @head.next_node
      while !ptr.equal?(@head)
        yield ptr
        ptr = ptr.next_node
      end
    end

    def empty?
      @no == 0
    end

    def search(data)
      ptr = @head.next_node
      while !ptr.equal?(@head)
        return ptr if ptr.data == data

        ptr = ptr.next_node
      end
      nil
    end

    def add_first(data)
      node = DLLNode.new(data, @head, @head.next_node)
      @head.next_node.prev_node = node
      @head.next_node = node
      @no += 1
    end

    def add_last(data)
      node = DLLNode.new(data, @head.prev_node, @head)
      @head.prev_node.next_node = node
      @head.prev_node = node
      @no += 1
    end

    def remove(node)
      return if empty?

      node.prev_node.next_node = node.next_node
      node.next_node.prev_node = node.prev_node
      @no -= 1
    end

    def clear
      @head.prev_node = @head
      @head.next_node = @head
      @no = 0
    end
  end

  NULL = -1

  # 線形リストノードクラス（配列カーソル版）
  class ArrayNode
    attr_accessor :data, :next_idx, :dnext

    def initialize(data = NULL, next_idx = NULL, dnext = NULL)
      @data = data
      @next_idx = next_idx
      @dnext = dnext
    end
  end

  # 線形リストクラス（配列カーソル版）
  class ArrayLinkedList
    def initialize(capacity)
      @head = NULL
      @current = NULL
      @max = NULL
      @deleted = NULL
      @capacity = capacity
      @n = Array.new(@capacity) { ArrayNode.new }
      @no = 0
    end

    def length
      @no
    end

    def get_insert_index
      if @deleted == NULL
        if @max + 1 < @capacity
          @max += 1
          return @max
        else
          return NULL
        end
      else
        rec = @deleted
        @deleted = @n[rec].dnext
        rec
      end
    end

    def add_first(data)
      ptr = @head
      rec = get_insert_index
      if rec != NULL
        @head = @current = rec
        @n[@head] = ArrayNode.new(data, ptr)
        @no += 1
      end
    end

    def add_last(data)
      if @head == NULL
        add_first(data)
      else
        ptr = @head
        ptr = @n[ptr].next_idx while @n[ptr].next_idx != NULL
        rec = get_insert_index
        if rec != NULL
          @n[ptr].next_idx = @current = rec
          @n[rec] = ArrayNode.new(data)
          @no += 1
        end
      end
    end

    def search(data)
      ptr = @head
      while ptr != NULL
        if @n[ptr].data == data
          @current = ptr
          return ptr
        end
        ptr = @n[ptr].next_idx
      end
      NULL
    end

    def remove_first
      if @head != NULL
        ptr = @head
        @head = @current = @n[ptr].next_idx
        @n[ptr].dnext = @deleted
        @deleted = ptr
        @no -= 1
      end
    end
  end
end
